functional_tests/src/runtest/main.go

```
@@ -0,0 +1,191 @@
+	}
+
+	info("%s>>%s<<%s", name, content, name)
+}
+
+func doSetup(suite Suite) {
+	info(">>> Setting up...")
+	for _, setup := range suite.Setup {
+		err, _, _ := runCommand(setup.Command[:1][0], setup.Command[1:]...)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+}
+
+func doTeardown(suite Suite) {
+	info(">>> Tearing down...")
+	for _, teardown := range suite.Teardown {
+		err, _, _ := runCommand(teardown.Command[:1][0], teardown.Command[1:]...)
+		if err != nil {
+			info(" > Warning: Teardown command failed: %s", err)
+		}
+	}
+}
+
+func doRunTests(suite Suite) func() {
+	info(">>> Testing...")
+	results := make(map[int]string)
+	passing := make(map[int]bool)
+	for index, test := range suite.Tests {
+		doOneTest(index, test, suite, results, passing)
+	}
+
+	report := func() {
+		info(">>> Summary:")
+		for index := range suite.Tests {
+			info(" > %s", results[index])
+		}
+
+		for index := range suite.Tests {
+			if !passing[index] {
+				log.Fatalf("Some test has failed.")
+			}
+		}
+	}
+	return report
+}
+
+func doOneTest(index int, test Test, suite Suite, results map[int]string, passing map[int]bool) {
+	var name string
+	var msg string
+	var result string
+
+	if len(test.Name) > 0 {
+		name = fmt.Sprintf("test-%d (%s)", index, test.Name)
+	} else {
+		name = fmt.Sprintf("test-%d", index)
+	}
+	info(" > %s", name)
+
+	recordResult := func(b *string) {
+		if *b == PASSED {
+			results[index] = fmt.Sprintf("%s: %s", name, PASSED)
+			passing[index] = true
+		} else {
+			results[index] = fmt.Sprintf("%s: %s", name, FAILED)
+			passing[index] = false
+		}
+		info(" %s", *b)
+	}
+	defer recordResult(&result)
+
+	args := append([]string{"exec", suite.Target}, test.Command...)
+	err, stdout, stderr := runCommand("docker", args...)
+
+	if err != nil {
+		if exiterr, ok := err.(*exec.ExitError); ok {
+			if status, ok := exiterr.Sys().(syscall.WaitStatus); ok {
+				result = fmt.Sprintf("%s: Exit status %d", FAILED, status.ExitStatus())
+			}
+		} else {
```

If the command fails to *start* (instead of failing to *finish*), the error returned will be of type [exec.Error](https://golang.org/pkg/os/exec/#Error), which has additional information that you could make use of.